pkg/operator/upgradebackupcontroller/upgradebackupcontroller.go

```diff
@@ -0,0 +1,258 @@
+			Type:   "RecentBackup",
+			Status: operatorv1.ConditionUnknown,
+			Reason: backupVersion,
+		}))
+		if err != nil {
+			return err
+		}
+	} else if backupCondition.Status != operatorv1.ConditionUnknown {
+		// Backup was taken or failed, but is not in progress.
+		return nil
+	}
+
+	backupStatus, err := ensureRecentBackup(ctx, c.kubeClient, c.podLister, c.nodeLister, backupVersion, c.targetImagePullSpec, syncCtx.Recorder())
+	if err != nil {
+		return err
+	}
+	if backupStatus == operatorv1.ConditionUnknown {
+		// Backup is still in progress
+		return nil
+	}
+
+	// Backup has completed or failed, update status accordingly
+	_, _, err = v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "RecentBackup",
+		Status: backupStatus,
+		Reason: backupVersion,
+	}))
+	return err
+}
+
+func ensureRecentBackup(ctx context.Context, kubeClient kubernetes.Interface, podLister corev1listers.PodLister, nodeLister corev1listers.NodeLister, clusterBackupVersion, targetImagePullSpec string, recorder events.Recorder) (operatorv1.ConditionStatus, error) {
+	delete := func(pod *corev1.Pod) error {
+		err := kubeClient.CoreV1().Pods(operatorclient.TargetNamespace).Delete(ctx, pod.Name, metav1.DeleteOptions{Preconditions: metav1.NewUIDPreconditions(string(pod.UID))})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	backupPods, err := podLister.List(labels.Set{"app": "cluster-backup"}.AsSelector())
+	if err != nil {
+		return "", err
+	}
+
+	var errs []error
+	for _, pod := range backupPods {
+		if pod.Labels[clusterBackupVersionLabel] != clusterBackupVersion {
+			// Delete pods with invalid version
+			if err := delete(pod); err != nil {
+				errs = append(errs, err)
+			}
+			continue
+		}
+
+		switch pod.Status.Phase {
+		case corev1.PodSucceeded:
+			return operatorv1.ConditionTrue, nil
+		case corev1.PodFailed:
+			if err := delete(pod); err != nil {
+				klog.Errorf("failed to delete pod %s/%s: %v", pod.Namespace, pod.Name, err)
+			}
+			return operatorv1.ConditionFalse, nil
+		default:
+			return operatorv1.ConditionUnknown, nil
+		}
+	}
+
+	node, err := getNodeWithAdequateAllocatableStorage(ctx, podLister, nodeLister)
+	if err != nil {
+		return "", err
+	}
+
+	_, _, err = manageRecoveryPod(ctx, node.Name, targetImagePullSpec, clusterBackupVersion, kubeClient.CoreV1(), recorder)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("%q: %v", "pod/cluster-backup", err))
+	}
+
+	if len(errs) > 0 {
+		return "", utilerrors.NewAggregate(errs)
+	}
+
+	return operatorv1.ConditionUnknown, nil
+}
+
+func manageRecoveryPod(ctx context.Context, nodeName, targetImagePullSpec, clusterBackupVersion string, client coreclientv1.PodsGetter, recorder events.Recorder) (*corev1.Pod, bool, error) {
+	pod := resourceread.ReadPodV1OrDie(etcd_assets.MustAsset("etcd/cluster-backup-pod.yaml"))
+	pod.Name = "cluster-backup-" + nodeName
```

Comment:
Do we need `nodeName` here?  In `ensureRecentBackup` you were already failing fast if there were existing backup pods, so I'm having trouble seeing a path to having multiple going at once.  If we do feel like there could be multiple going at once, I don't see any guards to try to keep them on different nodes.  Can we use a random suffix or some such here if we want to support parallel backup pods, or drop the suffix entirely if we don't want to support parallel backup pods?

Reply:
yeah +1 I think this is the reality I used an initial design that was more robust but simplicity is the key and I will update to reflect.